File: go/cmd/aontu/render_matrix_test.go
/* Copyright (c) 2026 Richard Rodger, MIT License */

package main

import (
	"encoding/json"
	"os"
	"path/filepath"
	"regexp"
	"runtime"
	"sort"
	"strconv"
	"strings"
	"testing"
)

// The render matrix: test/render/cases.json, held to the same
// expectations by ts/test/render-matrix.test.ts. A change to one port's
// render output fails the other port too.

type matrixStep struct {
	Args   []string          `json:"args"`
	Code   *int              `json:"code"`
	Out    *string           `json:"out"`
	Err    *string           `json:"err"`
	Edit   map[string]string `json:"edit"`
	Remove []string          `json:"remove"`
	Chmod  map[string]string `json:"chmod"`
}

type matrixCase struct {
	Name      string            `json:"name"`
	Posix     bool              `json:"posix"`
	Files     map[string]string `json:"files"`
	Steps     []matrixStep      `json:"steps"`
	After     map[string]string `json:"after"`
	AfterMode map[string]string `json:"afterMode"`
}

// The runtime's own record is excluded: its log carries the time of the
// run, and nothing else here does.
func matrixHeld(t *testing.T, dir string) []string {
	t.Helper()
	out := []string{}
	err := filepath.WalkDir(dir, func(at string, e os.DirEntry, err error) error {
		if nil != err {
			return err
		}
		if e.IsDir() {
			if ".jostraca" == e.Name() {
				return filepath.SkipDir
			}
			return nil
		}
		rel, err := filepath.Rel(dir, at)
		if nil != err {
			return err
		}
		out = append(out, filepath.ToSlash(rel))
		return nil
	})
	if nil != err {
		t.Fatal(err)
	}
	sort.Strings(out)
	return out
}

var matrixVersion = regexp.MustCompile(`"version": "[^"]*"`)

func matrixNorm(text, dir string) string {
	slashed := strings.ReplaceAll(text, "\\", "/")
	slashed = strings.ReplaceAll(slashed, strings.ReplaceAll(dir, "\\", "/"), "{dir}")
	return matrixVersion.ReplaceAllString(slashed, `"version": "{version}"`)
}

func matrixWrite(t *testing.T, dir string, files map[string]string) {
	t.Helper()
	for rel, content := range files {
		at := filepath.Join(dir, filepath.FromSlash(rel))
		if err := os.MkdirAll(filepath.Dir(at), 0o755); nil != err {
			t.Fatal(err)
		}
		if err := os.WriteFile(at, []byte(content), 0o600); nil != err {
			t.Fatal(err)
		}
	}
}

func TestRenderMatrix(t *testing.T) {
	raw, err := os.ReadFile(filepath.Join("..", "..", "..", "test", "render", "cases.json"))
	if nil != err {
		t.Fatal(err)
	}
	var file struct {
		Cases []matrixCase `json:"cases"`
	}
	if err := json.Unmarshal(raw, &file); nil != err {
		t.Fatal(err)
	}
	if 0 == len(file.Cases) {
		t.Fatal("the render matrix is empty")
	}

	for _, c := range file.Cases {
		t.Run(c.Name, func(t *testing.T) {
			if c.Posix && "windows" == runtime.GOOS {
				t.Skip("permission bits mean nothing here")
			}
			dir, err := filepath.EvalSymlinks(t.TempDir())
			if nil != err {
				t.Fatal(err)
			}
			matrixWrite(t, dir, c.Files)

			for n, step := range c.Steps {
				if nil != step.Edit || nil != step.Remove || nil != step.Chmod {
					matrixWrite(t, dir, step.Edit)
					for _, rel := range step.Remove {
						if err := os.Remove(filepath.Join(dir, filepath.FromSlash(rel))); nil != err {
							t.Fatal(err)
						}
					}
					for rel, mode := range step.Chmod {
						bits, err := strconv.ParseUint(mode, 8, 32)
						if nil != err {
							t.Fatal(err)
						}
						if err := os.Chmod(filepath.Join(dir, filepath.FromSlash(rel)),
							os.FileMode(bits)); nil != err {
							t.Fatal(err)
						}
					}
					continue
				}
				args := make([]string, 0, len(step.Args))
				for _, a := range step.Args {
					args = append(args, strings.ReplaceAll(a, "{dir}", dir))
				}
				out, errw, code := renderRunCLI(args...)
				at := c.Name + " step " + string(rune('1'+n))
				if nil != step.Out && matrixNorm(out, dir) != *step.Out {
					t.Fatalf("%s stdout: %q, want %q", at, matrixNorm(out, dir), *step.Out)
				}
				if nil != step.Err && matrixNorm(errw, dir) != *step.Err {
					t.Fatalf("%s stderr: %q, want %q", at, matrixNorm(errw, dir), *step.Err)
				}
				if nil != step.Code && code != *step.Code {
					t.Fatalf("%s exit code: %d, want %d", at, code, *step.Code)
				}
			}

			held := []string{}
			for _, rel := range matrixHeld(t, dir) {
				if _, input := c.Files[rel]; !input {
					held = append(held, rel)
				}
			}
			want := make([]string, 0, len(c.After))
			for rel := range c.After {
				want = append(want, rel)
			}
			sort.Strings(want)
			if strings.Join(held, "\n") != strings.Join(want, "\n") {
				t.Fatalf("%s: the tree holds %v, want %v", c.Name, held, want)
			}

			for rel, want := range c.After {
				got := "<absent>"
				if b, err := os.ReadFile(filepath.Join(dir, filepath.FromSlash(rel))); nil == err {
					got = string(b)
				}
				if got != want {
					t.Fatalf("%s: %s is %q, want %q", c.Name, rel, got, want)
				}
			}

			for rel, want := range c.AfterMode {
				st, err := os.Stat(filepath.Join(dir, filepath.FromSlash(rel)))
				if nil != err {
					t.Fatal(err)
				}
				if got := strconv.FormatUint(uint64(st.Mode().Perm()), 8); got != want {
					t.Fatalf("%s: %s mode is %s, want %s", c.Name, rel, got, want)
				}
			}
		})
	}
}

File: ts/test/render-matrix.test.ts
/* Copyright (c) 2026 Richard Rodger, MIT License */

// The render matrix: test/render/cases.json, held to the same
// expectations by go/cmd/aontu/render_matrix_test.go. A change to one
// port's render output fails the other port too.

import { describe, test } from 'node:test'
import * as Assert from 'node:assert'
import * as Fs from 'node:fs'
import * as Os from 'node:os'
import * as Path from 'node:path'

import { runRender } from '../dist/cli'

const CASES = Path.join(__dirname, '..', '..', 'test', 'render', 'cases.json')

type Step = {
  args?: string[], code?: number, out?: string, err?: string,
  edit?: { [path: string]: string }, remove?: string[],
}

type Case = {
  name: string,
  files: { [path: string]: string },
  steps: Step[],
  after?: { [path: string]: string },
}


function norm(text: string, dir: string): string {
  const slashed = text.split('\\').join('/')
  return slashed.split(dir.split('\\').join('/')).join('{dir}')
    .replace(/"version": "[^"]*"/g, '"version": "{version}"')
}


function write(dir: string, files?: { [path: string]: string }) {
  for (const [rel, content] of Object.entries(files ?? {})) {
    const at = Path.join(dir, rel)
    Fs.mkdirSync(Path.dirname(at), { recursive: true })
    Fs.writeFileSync(at, content)
  }
}


async function capture(args: string[]) {
  const so = process.stdout.write
  const se = process.stderr.write
  let out = ''
  let err = ''
  ;(process.stdout as any).write = (s: any) => ((out += s), true)
  ;(process.stderr as any).write = (s: any) => ((err += s), true)
  try {
    const code = await runRender(args)
    return { out, err, code }
  }
  finally {
    process.stdout.write = so
    process.stderr.write = se
  }
}


describe('render-matrix', () => {

  const cases: Case[] = JSON.parse(Fs.readFileSync(CASES, 'utf8')).cases
  Assert.ok(0 < cases.length, 'the render matrix is empty')

  for (const c of cases) {
    test(c.name, async () => {
      const dir = Fs.realpathSync(Fs.mkdtempSync(Path.join(Os.tmpdir(), 'aontu-matrix-')))
      write(dir, c.files)

      let n = 0
      for (const step of c.steps) {
        n++
        if (null != step.edit || null != step.remove) {
          write(dir, step.edit)
          for (const rel of step.remove ?? []) {
            Fs.rmSync(Path.join(dir, rel))
          }
          continue
        }
        const args = (step.args ?? []).map((a) => a.split('{dir}').join(dir))
        const got = await capture(args)
        const at = `${c.name} step ${n}`
        if (null != step.out) Assert.equal(norm(got.out, dir), step.out, at + ' stdout')
        if (null != step.err) Assert.equal(norm(got.err, dir), step.err, at + ' stderr')
        if (null != step.code) Assert.equal(got.code, step.code, at + ' exit code')
      }

      for (const [rel, want] of Object.entries(c.after ?? {})) {
        const at = Path.join(dir, rel)
        Assert.equal(Fs.existsSync(at) ? Fs.readFileSync(at, 'utf8') : '<absent>',
          want, `${c.name}: ${rel}`)
      }
    })
  }
})
